sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Pipeline
+{
+    internal static class TaskExtensions
+    {
+        public static T EnsureCompleted<T>(this Task<T> task)
+        {
+#if DEBUG
+            VerifyTaskCompleted(task.IsCompleted);
+#endif
+#pragma warning disable AZC0102
+            return task.GetAwaiter().GetResult();
+#pragma warning restore AZC0102
+        }
+
+        public static void EnsureCompleted(this Task task)
+        {
+#if DEBUG
+            VerifyTaskCompleted(task.IsCompleted);
+#endif
+#pragma warning disable AZC0102
+            task.GetAwaiter().GetResult();
+#pragma warning restore AZC0102
+        }
+
+        public static T EnsureCompleted<T>(this ValueTask<T> task)
+        {
+#if DEBUG
+            VerifyTaskCompleted(task.IsCompleted);
+#endif
+#pragma warning disable AZC0102
+            return task.GetAwaiter().GetResult();
+#pragma warning restore AZC0102
+        }
+
+        public static void EnsureCompleted(this ValueTask task)
+        {
+#if DEBUG
+            VerifyTaskCompleted(task.IsCompleted);
+#endif
+#pragma warning disable AZC0102
+            task.GetAwaiter().GetResult();
+#pragma warning restore AZC0102
+        }
+
+        public static ConfiguredValueTaskAwaitable<T> EnsureCompleted<T>(this ConfiguredValueTaskAwaitable<T> task, bool async)
+        {
+#if DEBUG
+            if (!async)
+            {
+                VerifyTaskCompleted(task.GetAwaiter().IsCompleted);
+            }
+#endif
+            return task;
+        }
+
+        public static ConfiguredValueTaskAwaitable EnsureCompleted(this ConfiguredValueTaskAwaitable task, bool async)
+        {
+#if DEBUG
+            if (!async)
+            {
+                VerifyTaskCompleted(task.GetAwaiter().IsCompleted);
+            }
+#endif
+            return task;
+        }
+
+        [Conditional("DEBUG")]
+        private static void VerifyTaskCompleted(bool isCompleted)
+        {
+            if (!isCompleted)
+            {
+                if (Debugger.IsAttached)
+                {
+                    Debugger.Break();
+                }
+                else
+                {
+                    // Throw an InvalidOperationException instead of using
+                    // Debug.Assert because that brings down nUnit immediately
+                    throw new InvalidOperationException("Task is not completed");
+                }
```

Comment:
I don't see much value in that. We don't expect to handle this `InvalidOperationException` anywhere (cause it is `DEBUG`-only), so it will just bring down the app. You can just stop the debugging here.

Reply:
Is there any reason folks should continue debugging at this point?  I like the exception in case they F5 past the break and don't realize they've entered into a world of pain by skipping over the root of the failure.

Reply:
Well, I don't have any strong preference here. Updated to always throwing exception.